File: iosApp/iosApp/blockchain/structs/OfferStruct.swift
//
//  OfferStruct.swift
//  iosApp
//
//  Created by jimmyt on 6/24/22.
//  Copyright © 2022 orgName. All rights reserved.
//

import BigInt
import Foundation
import web3swift

/**
 Represents an on-chain [Offer](https://www.commuto.xyz/docs/technical-reference/core-tec-ref#offer) struct.
 */
struct OfferStruct {
    /**
     Corresponds to an on-chain Offer's `isCreated` property.
     */
    let isCreated: Bool
    /**
     Corresponds to an on-chain Offer's `isTaken` property.
     */
    let isTaken: Bool
    /**
     Corresponds to an on-chain Offer's `maker` property.
     */
    let maker: EthereumAddress
    /**
     Corresponds to an on-chain Offer's `interfaceId` property.
     */
    let interfaceId: Data
    /**
     Corresponds to an on-chain Offer's `stablecoin` property.
     */
    let stablecoin: EthereumAddress
    /**
     Corresponds to an on-chain Offer's `amountLowerBound` property.
     */
    let amountLowerBound: BigUInt
    /**
     Corresponds to an on-chain Offer's `amountUpperBound` property.
     */
    let amountUpperBound: BigUInt
    /**
     Corresponds to an on-chain Offer's `securityDepositAmount` property.
     */
    let securityDepositAmount: BigUInt
    /**
     Corresponds to an on-chain Offer's `serviceFeeRate` property.
     */
    let serviceFeeRate: BigUInt
    /**
     Corresponds to an on-chain Offer's `direction` property.
     */
    let direction: BigUInt
    /**
     Corresponds to an on-chain Offer's `settlementMethods` property.
     */
    let settlementMethods: [Data]
    /**
     Corresponds to an on-chain Offer's `protocolVersion` property.
     */
    let protocolVersion: BigUInt
    /**
     The ID of the blockchain on which this Offer exists.
     */
    let chainID: BigUInt
    
    /**
     Creates an `OfferStruct` from the `[AnyObject]` obtained via a web3swift read transaction call to [getOffer](https://www.commuto.xyz/docs/technical-reference/core-tec-ref#get-offer).
     
     - Parameters:
        - response: The `[AnyObject]` obtained by a read transaction call to getOffer.
        - chainID: The ID of the blockchain from which `response` was obtained.
     
     - Returns: A new `OfferStruct` with property values derived from `response` and `chainID`.
     
     - Throws: `BlockchainServiceError.unexpectedNilError` if an expected parameter in `response` is missing.
     */
    static func createFromGetOfferResponse(response: [AnyObject], chainID: BigUInt) throws -> OfferStruct {
        guard let isCreated = response[0] as? Bool else {
            throw BlockchainServiceError.unexpectedNilError(desc: makeCreationErrorMessage(valueName: "isCreated"))
        }
        guard let isTaken: Bool = response[1] as? Bool else {
            throw BlockchainServiceError.unexpectedNilError(desc: makeCreationErrorMessage(valueName: "isTaken"))
        }
        guard let maker = response[2] as? EthereumAddress else {
            throw BlockchainServiceError.unexpectedNilError(desc: makeCreationErrorMessage(valueName: "maker"))
        }
        guard let interfaceId: Data = response[3] as? Data else {
            throw BlockchainServiceError.unexpectedNilError(desc: makeCreationErrorMessage(valueName: "interfaceId"))
        }
        guard let stablecoin = response[4] as? EthereumAddress else {
            throw BlockchainServiceError.unexpectedNilError(desc: makeCreationErrorMessage(valueName: "stablecoin"))
        }
        guard let amountLowerBound = response[5] as? BigUInt else {
            throw BlockchainServiceError.unexpectedNilError(desc: makeCreationErrorMessage(valueName: "amountLowerBound"))
        }
        guard let amountUpperBound = response[6] as? BigUInt else {
            throw BlockchainServiceError.unexpectedNilError(desc: makeCreationErrorMessage(valueName: "amountUpperBound"))
        }
        guard let securityDepositAmount = response[7] as? BigUInt else {
            throw BlockchainServiceError.unexpectedNilError(desc: makeCreationErrorMessage(valueName: "securityDepositAmount"))
        }
        guard let serviceFeeRate = response[8] as? BigUInt else {
            throw BlockchainServiceError.unexpectedNilError(desc: makeCreationErrorMessage(valueName: "serviceFeeRate"))
        }
        guard let direction = response[9] as? BigUInt else {
            throw BlockchainServiceError.unexpectedNilError(desc: makeCreationErrorMessage(valueName: "direction"))
        }
        guard let settlementMethods: [Data] = response[10] as? [Data] else {
            throw BlockchainServiceError.unexpectedNilError(desc: makeCreationErrorMessage(valueName: "settlementMethods"))
        }
        guard let protocolVersion = response[11] as? BigUInt else {
            throw BlockchainServiceError.unexpectedNilError(desc: makeCreationErrorMessage(valueName: "protocolVersion"))
        }
        return OfferStruct(isCreated: isCreated, isTaken: isTaken, maker: maker, interfaceId: interfaceId, stablecoin: stablecoin, amountLowerBound: amountLowerBound, amountUpperBound: amountUpperBound, securityDepositAmount: securityDepositAmount, serviceFeeRate: serviceFeeRate, direction: direction, settlementMethods: settlementMethods, protocolVersion: protocolVersion, chainID: chainID)
    }
    
    /**
     Called by `createFromGetOfferResponse` to get the appropriate message for a `BlockchainService.unexpectedNilError` when `createFromGetOfferResponse` is unable to get the value of a field that is expected in its `response` parameter.
     */
    private static func makeCreationErrorMessage(valueName: String) -> String {
        return "Got nil while getting " + valueName + " value from getOffer response"
    }
    
}

File: androidApp/src/main/java/com/commuto/interfacemobile/android/blockchain/structs/OfferStruct.kt
package com.commuto.interfacemobile.android.blockchain.structs

import com.commuto.interfacemobile.android.contractwrapper.CommutoSwap
import java.math.BigInteger

/**
 * Represents an on-chain [Offer](https://www.commuto.xyz/docs/technical-reference/core-tec-ref#offer) struct.
 *
 * @param isCreated Corresponds to an on-chain Offer's `isCreated` property.
 * @param isTaken Corresponds to an on-chain Offer's `isTaken` property.
 * @param maker Corresponds to an on-chain Offer's `maker` property.
 * @param interfaceID Corresponds ot an on-chain Offer's `interfaceId` property.
 * @param stablecoin Corresponds to an on-chain Offer's `stablecoin` property.
 * @param amountLowerBound Corresponds to an on-chain Offer's `amountLowerBound` property.
 * @param amountUpperBound Corresponds to an on-chain Offer's `amountUpperBound` property.
 * @param securityDepositAmount Corresponds to an on-chain Offer's `securityDepositAmount` property.
 * @param serviceFeeRate Corresponds to an on-chain Offer's `serviceFeeRate` property.
 * @param direction Corresponds to an on-chain Offer's `direction` property.
 * @param settlementMethods Corresponds to an on-chain Offer's `settlementMethods` property.
 * @param protocolVersion Corresponds to an on-chain Offer's `protocolVersion` property.
 * @param chainID Corresponds to an on-chain Offer's `chainID` property.
 */
data class OfferStruct(
    val isCreated: Boolean,
    val isTaken: Boolean,
    val maker: String,
    val interfaceID: ByteArray,
    val stablecoin: String,
    val amountLowerBound: BigInteger,
    val amountUpperBound: BigInteger,
    val securityDepositAmount: BigInteger,
    val serviceFeeRate: BigInteger,
    val direction: BigInteger,
    val settlementMethods: List<ByteArray>,
    val protocolVersion: BigInteger,
    val chainID: BigInteger
) {
    /**
     * Creates an [OfferStruct] from a [CommutoSwap.Offer] and the specified [chainID].
     *
     * @param offer The [CommutoSwap.Offer] from which the returned [OfferStruct] will be created.
     * @param chainID The ID of the blockchain on which this offer exists.
     */
    private constructor(offer: CommutoSwap.Offer, chainID: BigInteger): this(
        offer.isCreated,
        offer.isTaken,
        offer.maker,
        offer.interfaceId,
        offer.stablecoin,
        offer.amountLowerBound,
        offer.amountUpperBound,
        offer.securityDepositAmount,
        offer.serviceFeeRate,
        offer.direction,
        offer.settlementMethods,
        offer.protocolVersion,
        chainID
    )

    /**
     * Returns a [CommutoSwap.Offer] derived from this [OfferStruct].
     *
     * @return An [CommutoSwap.Offer] derived from this [OfferStruct].
     */
    fun toCommutoSwapOffer(): CommutoSwap.Offer {
        return CommutoSwap.Offer(
            this.isCreated,
            this.isTaken,
            this.maker,
            this.interfaceID,
            this.stablecoin,
            this.amountLowerBound,
            this.amountUpperBound,
            this.securityDepositAmount,
            this.serviceFeeRate,
            this.direction,
            this.settlementMethods,
            this.protocolVersion
        )
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as OfferStruct

        if (isCreated != other.isCreated) return false
        if (isTaken != other.isTaken) return false
        if (maker != other.maker) return false
        if (!interfaceID.contentEquals(other.interfaceID)) return false
        if (stablecoin != other.stablecoin) return false
        if (amountLowerBound != other.amountLowerBound) return false
        if (amountUpperBound != other.amountUpperBound) return false
        if (securityDepositAmount != other.securityDepositAmount) return false
        if (serviceFeeRate != other.serviceFeeRate) return false
        if (direction != other.direction) return false
        if (settlementMethods != other.settlementMethods) return false
        if (protocolVersion != other.protocolVersion) return false
        if (chainID != other.chainID) return false

        return true
    }

    override fun hashCode(): Int {
        var result = isCreated.hashCode()
        result = 31 * result + isTaken.hashCode()
        result = 31 * result + maker.hashCode()
        result = 31 * result + interfaceID.contentHashCode()
        result = 31 * result + stablecoin.hashCode()
        result = 31 * result + amountLowerBound.hashCode()
        result = 31 * result + amountUpperBound.hashCode()
        result = 31 * result + securityDepositAmount.hashCode()
        result = 31 * result + serviceFeeRate.hashCode()
        result = 31 * result + direction.hashCode()
        result = 31 * result + settlementMethods.hashCode()
        result = 31 * result + protocolVersion.hashCode()
        result = 31 * result + chainID.hashCode()
        return result
    }

    companion object {
        /**
         * Creates an [OfferStruct] from a [CommutoSwap.Offer] and the specified [chainID] if
         * [CommutoSwap.Offer.isCreated] is true, or returns null if false.
         *
         * @param offer The [CommutoSwap.Offer] from which the returned [OfferStruct] will be created.
         * @param chainID The ID of the blockchain on which this offer exists.
         *
         * @return An [OfferStruct] derived from [offer] and [chainID] if [CommutoSwap.Offer.isCreated] is true, or null
         * if false.
         */
        fun createFromGetOfferResponse(offer: CommutoSwap.Offer, chainID: BigInteger): OfferStruct? {
            return if (!offer.isCreated) {
                null
            } else {
                OfferStruct(offer, chainID)
            }
        }
    }
}